core/consensus/src/adapter.rs

```diff
@@ -387,6 +444,178 @@ where
         self.mempool
             .set_args(timeout_gap, cycles_limit, max_tx_size);
     }
+
+    /// make sure that the mempool is in the same height!
+    async fn verify_txs(&self, ctx: Context, txs: Vec<Hash>) -> ProtocolResult<()> {
+        self.mempool.ensure_order_txs(ctx.clone(), txs).await
+    }
+
+    /// this function verify all info in header except proof and roots
+    async fn verify_block_header(&self, ctx: Context, block: &Block) -> ProtocolResult<()> {
+        let previous_block = self
+            .get_block_by_height(ctx.clone(), block.header.height - 1)
+            .await?;
+
+        let previous_block_hash = Hash::digest(previous_block.encode_fixed()?);
+
+        if previous_block_hash != block.header.pre_hash {
+            return Err(
+                ConsensusError::VerifyBlockHeaderPreBlockHashErr(block.header.height).into(),
+            );
+        }
+
+        // the block 0 and 1 's proof is consensus-ed by community
+        if block.header.pre_hash != block.header.proof.block_hash && block.header.height > 1 {
+            log::error!("verifying_block : {:?}", block);
+            return Err(ConsensusError::VerifyBlockHeaderPreHashErr(block.header.height).into());
+        }
+
+        // verify proposer and validators
+        let previous_metadata = self.get_metadata(
+            ctx,
+            previous_block.header.state_root.clone(),
+            previous_block.header.height,
+            previous_block.header.timestamp,
+        )?;
+
+        let previous_authority_list = previous_metadata
+            .verifier_list
+            .iter()
+            .map(|v| Node {
+                address:        v.address.as_bytes(),
+                propose_weight: v.propose_weight,
+                vote_weight:    v.vote_weight,
+            })
+            .collect::<Vec<Node>>();
+
+        let mut authority_map = HashMap::new();
+        for node in previous_authority_list.iter() {
+            authority_map.insert(node.address.clone(), node.clone());
+        }
```

Comment:
```suggestion
        let authority_map = previous_metadata
            .verifier_list
            .into_iter()
            .map(|v| {
                let address = v.address.as_bytes();
                let node = Node {
                    address:        v.address.as_bytes(),
                    propose_weight: v.propose_weight,
                    vote_weight:    v.vote_weight,
                };
                (address, node)
            })
            .collect::<HashMap<_, _>>();
```

Reply:
fantastic, this is what I need for process streams...